pkg/services/baremetal/host/host.go

```diff
@@ -1202,6 +1205,131 @@ func (s *Service) actionImageInstalling(ctx context.Context) actionResult {
 	}
 }
 
+func (s *Service) actionImageInstallingCustomImageURLCommand(ctx context.Context, sshClient sshclient.Client) actionResult {
+	host := s.scope.HetznerBareMetalHost
+
+	state, logFile, err := sshClient.StateOfImageURLCommand()
+	if err != nil {
+		return actionError{err: fmt.Errorf("StateOfImageURLCommand failed: %w", err)}
+	}
+
+	duration := time.Since(host.Spec.Status.LastUpdated.Time)
+	// Please keep the number (7) in sync with the docstring of ImageURL.
+	if duration > 7*time.Minute {
+		// timeout. Something has failed.
+		msg := fmt.Sprintf("ImageURLCommand timed out after %s. Deleting machine",
+			duration.Round(time.Second).String())
+		s.scope.Logger.Error(nil, msg, "logFile", logFile)
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			"ImageURLCommandTimedout", clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return s.recordActionFailure(infrav1.FatalError, msg)
+	}
+
+	switch state {
+	case sshclient.ImageURLCommandStateRunning:
+		return actionContinue{delay: 10 * time.Second}
+
+	case sshclient.ImageURLCommandStateFinishedSuccessfully:
+		record.Event(s.scope.HetznerBareMetalHost, "ImageURLCommandOutput", logFile)
+		s.scope.Logger.Info("ImageURLCommandOutput", "logFile", logFile)
+
+		// Update name in robot API
+		if _, err := s.scope.RobotClient.SetBMServerName(s.scope.HetznerBareMetalHost.Spec.ServerID, s.scope.Hostname()); err != nil {
+			record.Warn(s.scope.HetznerBareMetalHost, "SetBMServerNameFailed", err.Error())
+			s.handleRobotRateLimitExceeded(err, "SetBMServerName")
+			return actionError{err: fmt.Errorf("failed to update name of host in robot API: %w", err)}
+		}
+
+		if err := sshClient.Reboot().Err; err != nil {
+			err = fmt.Errorf("failed to reboot server (after install-image): %w", err)
+			record.Warn(s.scope.HetznerBareMetalHost, "RebootFailed", err.Error())
+			return actionError{err: err}
+		}
+
+		msg := "machine image and cloud-init data got installed (via image-url-command)"
+		createSSHRebootEvent(ctx, s.scope.HetznerBareMetalHost, msg)
+
+		// clear potential errors - all done
+		s.scope.HetznerBareMetalHost.ClearError()
+		return actionComplete{}
+
+	case sshclient.ImageURLCommandStateFailed:
+		record.Warn(s.scope.HetznerBareMetalHost, "InstallImageNotSuccessful", logFile)
+		msg := "image-url-command failed"
+		s.scope.Logger.Error(nil, msg, "logFile", logFile)
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			"ImageURLCommandFailed", clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return s.recordActionFailure(infrav1.FatalError, msg)
+
+	case sshclient.ImageURLCommandStateNotStarted:
+		data, err := s.scope.GetRawBootstrapData(ctx)
+		if err != nil {
+			return actionError{err: fmt.Errorf("baremetal GetRawBootstrapData failed: %w", err)}
+		}
+		if s.scope.ImageURLCommand == "" {
```

Comment:
is this the first time we check it for bare metal? If so, then we should not return an error but "nil" because we should stop requeuing. I'm not sure if we have another similar case already, because we have to find a return value that makes sense and that let's us stop reconciliation.